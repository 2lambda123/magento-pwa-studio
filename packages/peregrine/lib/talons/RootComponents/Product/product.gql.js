import { gql } from '@apollo/client';

import { ProductDetailsFragment } from './productDetailFragment.gql';

export const GET_STORE_CONFIG_DATA = gql`
    query getStoreConfigData {
        # eslint-disable-next-line @graphql-eslint/require-id-when-available
        storeConfig {
            store_code
            product_url_suffix
        }
    }
`;

export const GET_SINGLE_PRODUCT_DETAIL_QUERY = gql`
    query getProductDetailForProductPage($sku: String!) {
        products(filter: { sku: { eq: $sku } }) {
            items {
                id
                uid
                ...ProductDetailsFragment
            }
        }
    }
    ${ProductDetailsFragment}
`;

export default {
    getStoreConfigData: GET_STORE_CONFIG_DATA,
    getSingleProductDetailQuery: GET_SINGLE_PRODUCT_DETAIL_QUERY
};
