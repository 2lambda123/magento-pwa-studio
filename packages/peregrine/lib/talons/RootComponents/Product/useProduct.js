import { useQuery } from '@apollo/client';
import { useMemo } from 'react';

/**
 * A [React Hook]{@link https://reactjs.org/docs/hooks-intro.html} that
 * controls the logic for the Product Root Component.
 *
 * @kind function
 *
 * @param {object}      props
 * @param {Function}    props.mapProduct - A function for updating products to the proper shape.
 * @param {GraphQLAST}  props.queries.getProductQuery - Fetches product using a server query
 * @param {String}      props.urlKey - The url_key of this product.
 *
 * @returns {object}    result
 * @returns {Bool}      result.error - Indicates a network error occurred.
 * @returns {Bool}      result.loading - Indicates the query is in flight.
 * @returns {Bool}      result.product - The product's details.
 */
export const useProduct = props => {
    const { mapProduct, queries, urlKey } = props;

    const { error, loading, data } = useQuery(queries.getProductQuery, {
        fetchPolicy: 'cache-and-network',
        nextFetchPolicy: 'cache-first',
        variables: {
            urlKey
        }
    });

    const product = useMemo(() => {
        // If a product is out of stock _and_ the backend specifies not to
        // display OOS items, the items array will be empty.
        if (data && data.products.items[0]) {
            return mapProduct(data.products.items[0]);
        }

        // The product isn't in the cache and we don't have a response from GraphQL yet.
        return null;
    }, [data, mapProduct]);

    return {
        error,
        loading,
        product
    };
};
