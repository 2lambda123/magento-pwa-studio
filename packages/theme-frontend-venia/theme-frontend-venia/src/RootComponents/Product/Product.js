import { Component, createElement } from 'react';
import PropTypes from 'prop-types';

import classify from 'src/classify';
import Carousel from 'src/components/ProductImageCarousel';
import Options from 'src/components/ProductOptions';
import Quantity from 'src/components/ProductQuantity';
import RichText from 'src/components/RichText';
import mockData from './mockData';
import defaultClasses from './product.css';

class Product extends Component {
    static propTypes = {
        classes: PropTypes.shape({
            root: PropTypes.string
        }),
        data: PropTypes.shape({
            additionalInfo: PropTypes.string,
            description: PropTypes.string,
            images: PropTypes.arrayOf(
                PropTypes.shape({
                    id: PropTypes.string
                })
            ),
            name: PropTypes.string,
            options: PropTypes.arrayOf(
                PropTypes.shape({
                    id: PropTypes.string
                })
            ),
            price: PropTypes.string
        })
    };

    static defaultProps = {
        data: mockData
    };

    render() {
        const { classes, data } = this.props;

        return (
            <article className={classes.root}>
                <section className={classes.title}>
                    <h1 className={classes.productName}>
                        <span>{data.name}</span>
                    </h1>
                    <p className={classes.productPrice}>
                        <span>{data.price}</span>
                    </p>
                </section>
                <section className={classes.imageCarousel}>
                    <Carousel images={data.images} />
                </section>
                <section className={classes.actions}>
                    <button className={classes.action}>
                        <span>Add to Wishlist</span>
                    </button>
                </section>
                <section className={classes.options}>
                    <Options options={data.options} />
                </section>
                <section className={classes.quantity}>
                    <h2 className={classes.quantityTitle}>
                        <span>Quantity</span>
                    </h2>
                    <Quantity />
                </section>
                <section className={classes.cartActions}>
                    <button className={classes.addToCart}>
                        <span>Add to Cart</span>
                    </button>
                </section>
                <section className={classes.description}>
                    <h2 className={classes.descriptionTitle}>
                        <span>Product Description</span>
                    </h2>
                    <RichText content={data.description} />
                </section>
                <section className={classes.details}>
                    <h2 className={classes.detailsTitle}>
                        <span>Details</span>
                    </h2>
                    <RichText content={data.additionalInfo} />
                </section>
                <section className={classes.related}>
                    <h2 className={classes.relatedTitle}>
                        <span>Similar Products</span>
                    </h2>
                </section>
            </article>
        );
    }
}

export default classify(defaultClasses)(Product);
