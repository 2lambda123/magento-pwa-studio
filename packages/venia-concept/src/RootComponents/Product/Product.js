import React, { Component } from 'react';
import { bool, shape, number, arrayOf, string } from 'prop-types';

import { connect, Query } from 'src/drivers';
import { addItemToCart } from 'src/actions/cart';
import { loadingIndicator } from 'src/components/LoadingIndicator';
import ProductFullDetail from 'src/components/ProductFullDetail';
import getUrlKey from 'src/util/getUrlKey';
import productQuery from 'src/queries/getProductDetail.graphql';

/**
 * As of this writing, there is no single Product query type in the M2.3 schema.
 * The recommended solution is to use filter criteria on a Products query.
 * However, the `id` argument is not supported. See
 * https://github.com/magento/graphql-ce/issues/86
 * TODO: Replace with a single product query when possible.
 */
class Product extends Component {
    static propTypes = {
        data: shape({
            productDetail: shape({
                total_count: number,
                items: arrayOf(
                    shape({
                        id: number,
                        sku: string.isRequired,
                        price: shape({
                            regularPrice: shape({
                                amount: shape({
                                    currency: string.isRequired,
                                    value: number.isRequired
                                })
                            }).isRequired
                        }).isRequired,
                        image: string,
                        image_label: string,
                        media_gallery_entries: arrayOf(
                            shape({
                                label: string,
                                position: number.isRequired,
                                disabled: bool,
                                file: string.isRequired
                            })
                        ),
                        description: string,
                        short_description: string,
                        canonical_url: string
                    })
                ).isRequired
            }).isRequired
        })
    };

    addToCart = async (item, quantity) => {
        const { guestCartId } = this.props;
        await this.props.addItemToCart({ guestCartId, item, quantity });
    };

    componentDidMount() {
        window.scrollTo(0, 0);
    }

    render() {
        return (
            <Query
                query={productQuery}
                variables={{ urlKey: getUrlKey(), onServer: false }}
            >
                {({ loading, error, data }) => {
                    if (error) return <div>Data Fetch Error</div>;
                    if (loading) return loadingIndicator;

                    const product = data.productDetail.items[0];

                    return (
                        <ProductFullDetail
                            product={product}
                            addToCart={this.props.addItemToCart}
                        />
                    );
                }}
            </Query>
        );
    }
}

const mapDispatchToProps = {
    addItemToCart
};

export default connect(
    null,
    mapDispatchToProps
)(Product);
