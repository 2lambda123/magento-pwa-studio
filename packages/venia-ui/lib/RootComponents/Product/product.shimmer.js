import React, { useMemo } from 'react';
import { string } from 'prop-types';
import Shimmer from '../../components/Shimmer';
import { BreadcrumbShimmer } from '../../components/Breadcrumbs';
import defaultClasses from '../../components/ProductFullDetail/productFullDetail.css';
import { useStyle } from '../../classify';

const ProductShimmer = (props) => {
    const { productType } = props;
    const classes = useStyle(defaultClasses, props.classes);

    const options = useMemo(() => {
        if (productType.includes('Configurable')) {
            // TODO return different shimmer structure
            return (
                <Shimmer />
            );
        }

        return null;
    }, [productType]);

    return (
        <Fragment>
            <BreadcrumbShimmer />
            <div className={classes.root}>
                <section className={classes.title}>
                    <h1 className={classes.productName}>
                        {productDetails.name}
                    </h1>
                    <p className={classes.productPrice}>
                        <Price
                            currencyCode={productDetails.price.currency}
                            value={productDetails.price.value}
                        />
                    </p>
                </section>
                <section className={classes.imageCarousel}>
                    <Carousel images={mediaGalleryEntries} />
                </section>
                <FormError
                    classes={{
                        root: classes.formErrors
                    }}
                    errors={errors.get('form') || []}
                />
                <section className={classes.options}>{options}</section>
                <section className={classes.quantity}>
                    <h2 className={classes.quantityTitle}>
                        <FormattedMessage
                            id={'global.quantity'}
                            defaultMessage={'Quantity'}
                        />
                    </h2>
                    <QuantityFields
                        classes={{ root: classes.quantityRoot }}
                        min={1}
                        message={errors.get('quantity')}
                    />
                </section>
                <section className={classes.actions}>
                    {cartActionContent}
                    <Suspense fallback={null}>
                        <WishlistButton {...wishlistButtonProps} />
                    </Suspense>
                </section>
                <section className={classes.description}>
                    <h2 className={classes.descriptionTitle}>
                        <FormattedMessage
                            id={'productFullDetail.productDescription'}
                            defaultMessage={'Product Description'}
                        />
                    </h2>
                    <RichText content={productDetails.description} />
                </section>
                <section className={classes.details}>
                    <h2 className={classes.detailsTitle}>
                        <FormattedMessage
                            id={'global.sku'}
                            defaultMessage={'SKU'}
                        />
                    </h2>
                    <strong>{productDetails.sku}</strong>
                </section>
            </div>
        </Fragment>
    );
};

ProductShimmer.propTypes = {
    productType: string.required
};

export default ProductShimmer;
