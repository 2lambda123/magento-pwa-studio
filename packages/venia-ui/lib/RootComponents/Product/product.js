import React, { Fragment } from 'react';
import { FormattedMessage } from 'react-intl';
import { useProduct } from '@magento/peregrine/lib/talons/RootComponents/Product/useProduct';

import ErrorView from '@magento/venia-ui/lib/components/ErrorView';
import { Title, Meta } from '@magento/venia-ui/lib/components/Head';
import { fullPageLoadingIndicator } from '@magento/venia-ui/lib/components/LoadingIndicator';
import ProductFullDetail from '@magento/venia-ui/lib/components/ProductFullDetail';
import mapProduct from '@magento/venia-ui/lib/util/mapProduct';

/*
 * As of this writing, there is no single Product query type in the M2.3 schema.
 * The recommended solution is to use filter criteria on a Products query.
 * However, the `id` argument is not supported. See
 * https://github.com/magento/graphql-ce/issues/86
 * TODO: Replace with a single product query when possible.
 */

const Product = () => {
    const talonProps = useProduct({
        mapProduct
    });

    const { error, loading, product } = talonProps;

    if (loading && !product) return fullPageLoadingIndicator;
    if (error && !product) return <ErrorView />;
    if (!product) {
        return (
            <h1>
                <FormattedMessage
                    id={'product.outOfStockTryAgain'}
                    defaultMessage={
                        'This Product is currently out of stock. Please try again later.'
                    }
                />
            </h1>
        );
    }

    // Note: STORE_NAME is injected by Webpack at build time.
    const title = `${product.name} - ${STORE_NAME}`;

    return (
        <Fragment>
            <Title>{title}</Title>
            <Meta name="description" content={product.meta_description} />
            <ProductFullDetail product={product} />
        </Fragment>
    );
};

export default Product;
