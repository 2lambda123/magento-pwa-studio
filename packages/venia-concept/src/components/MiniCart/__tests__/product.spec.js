import React, { useMemo, useState } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import Product from '../product';

global.getComputedStyle = jest.fn().mockReturnValue({
    getPropertyValue: jest.fn().mockReturnValue('80px')
});
jest.mock('react', () => {
    const React = jest.requireActual('react');
    const memoSpy = jest.spyOn(React, 'useMemo');
    const stateSpy = jest.spyOn(React, 'useState');

    return Object.assign(React, {
        useMemo: memoSpy,
        useState: stateSpy
    });
});

const renderer = new ShallowRenderer();

const props = {
    beginEditItem: jest.fn(),
    currencyCode: 'US',
    item: {
        image: {
            file: 'unittest'
        },
        name: 'Unit Test Product',
        options: [],
        price: 99,
        qty: 1
    },
    removeItemFromCart: jest.fn()
};

test('it renders correctly', () => {
    const tree = renderer.render(<Product {...props} />);

    expect(tree).toMatchSnapshot();
});

test('it renders a mask div before the kebab if it is loading', () => {
    useState
        // isFavorite
        .mockReturnValueOnce([false, jest.fn()])
        // isLoading
        .mockReturnValueOnce([true, jest.fn()]);

    useMemo.mockReturnValueOnce('path/to/image/source');

    const tree = renderer.render(<Product {...props} />);

    expect(tree).toMatchSnapshot();
});
