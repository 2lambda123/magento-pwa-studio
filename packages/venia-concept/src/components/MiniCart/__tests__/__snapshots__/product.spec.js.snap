// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a mask div before the kebab if it is loading 1`] = `
<li>
  <img
    alt="Unit Test Product"
    src="path/to/image/source"
  />
  <div>
    Unit Test Product
  </div>
  <ProductOptions
    options={Array []}
  />
  <div>
    <div>
      <span>
        1
      </span>
      <span>
        ×
      </span>
      <span>
        <mockConstructor
          currencyCode="US"
          value={99}
        />
      </span>
    </div>
  </div>
  <div />
  <Kebab>
    <Section
      icon="Heart"
      isFilled={false}
      onClick={[Function]}
      text="Add to favorites"
    />
    <Section
      icon="Edit2"
      onClick={[Function]}
      text="Edit item"
    />
    <Section
      icon="Trash"
      onClick={[Function]}
      text="Remove item"
    />
  </Kebab>
</li>
`;

exports[`it renders correctly 1`] = `
<li>
  <img
    alt="Unit Test Product"
    src="/media/catalog/product/unittest?auto=webp&format=pjpg&width=80"
  />
  <div>
    Unit Test Product
  </div>
  <ProductOptions
    options={Array []}
  />
  <div>
    <div>
      <span>
        1
      </span>
      <span>
        ×
      </span>
      <span>
        <mockConstructor
          currencyCode="US"
          value={99}
        />
      </span>
    </div>
  </div>
  <Kebab>
    <Section
      icon="Heart"
      isFilled={false}
      onClick={[Function]}
      text="Add to favorites"
    />
    <Section
      icon="Edit2"
      onClick={[Function]}
      text="Edit item"
    />
    <Section
      icon="Trash"
      onClick={[Function]}
      text="Remove item"
    />
  </Kebab>
</li>
`;
